tensorflow_quantum/core/ops/circuit_execution_ops.py

```diff
@@ -173,7 +173,7 @@ def get_sampling_op(backend=None):
         tfq_utility_ops.padded_to_ragged(TFQWavefunctionSimulator.samples(
             programs, symbol_names, symbol_values, num_samples))
 
-    if isinstance(backend, (cirq.SimulatesSamples, cirq.Sampler)):
```

Comment:
`cirq.SimulatesSamples` has the same public interface as `cirq.Sampler`, and inherits `cirq.Sampler`, so I would argue that this is ok

Reply:
Oops, was totally confused with `SimulatesFinalState` there. You're right, this should get removed.

Reply:
I realize this relates to your other comment too.  Currently `batch_util` relies on being able to call the `simulate` method of the passed sampler, which is outside the contract of both `cirq.SimulatesSamples` and `cirq.Sampler`.  Turns out both the simulators we accept also indirectly inherit `cirq.Sampler`, so there should be a good way to push things into the batched sampler.

Reply:
Looking at this more, the structure of `batch_sample` bakes-in the assumption that it is best to keep the sampling for each circuit on its own thread, and make each thread blocking until the sampling for that circuit is done.  However, for `cirq.Sampler`s which call out to another machine or remote service, we would rather rely on the `run_async` method; then, in the current `batch_sample` structure, we would be limiting the number of calls to the number of threads. 
 Going to do some more thinking about this before I make a big update.

Reply:
This is precisely what the `quantum_context` PR #188 is for. Basically if we want to run against a sampler that isn't operating locally (i.e. is over gRPC or some other slow API where every call is very precious timewise) we would set a flag like `tfq.enable_engine_mode()` which would then trigger `cirq_ops` to use `engine_util.py` instead of `batch_util.py`. The big TODO here is to write `engine_util.py`. #188 just sets up the basic code to do the changeover between `batch_util` and `engine_util`.